File: src/Core/FinancialEvents/Services/ApplyIdentifierRelationshipsService.py
from typing import Sequence

import Core.FinancialEvents.Schemas.CommonFormats as cf
import Core.FinancialEvents.Schemas.Events as pe
import Core.FinancialEvents.Schemas.FinancialEvents as pfe
import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.FinancialEvents.Schemas.FinancialIdentifier import FinancialIdentifier
from Core.FinancialEvents.Schemas.IdentifierRelationship import (
    IdentifierChangeType,
    IdentifierRelationship,
    IdentifierRelationshipSplit,
)


class ApplyIdentifierRelationshipsService:
    """
    Bakes in selected identifier relationship types (e.g. RENAME, SPLIT, REVERSE_SPLIT)
    by producing new financial groupings.

    - RENAME: merges groupings along rename chains and normalizes to the sink identifier.
    - SPLIT / REVERSE_SPLIT: a corporate action changes the instrument identifier (e.g. ISIN).
      We treat the relationship as FromIdentifier (old ISIN) → ToIdentifier (new ISIN).
      Only the old-ISIN (From) grouping is scaled for that split; the new ISIN (To) is not.
      We scale pre-effective-date quantities in the From grouping, then merge it into the To
      grouping so all events end up under the new identifier. Applied in EffectiveDate order.
      In a chain (e.g. ISIN.OLD.OLD → split → ISIN.OLD → split → ISIN), each split is applied
      in order: the oldest ISIN is scaled and merged into the next, then that grouping is
      scaled by the next split and merged again, so the original position is scaled by every
      ratio along the chain.

    Tax authorities (or the pipeline) decide which change types to apply.

    Application order: RENAME first, then SPLIT/REVERSE_SPLIT by EffectiveDate.
    We apply renames before splits so that all same-instrument groupings (e.g. same ISIN,
    different ticker merged by a rename) are consolidated into one before we apply a split.
    Otherwise a split would only scale and merge the first matching "old ISIN" grouping
    (next()), leaving other same-ISIN groupings (e.g. other ticker) behind.
    """

    def apply(
        self,
        events: pfe.FinancialEvents,
        changeTypesToApply: Sequence[IdentifierChangeType],
    ) -> pfe.FinancialEvents:
        """Apply selected relationship types: RENAME first, then SPLIT/REVERSE_SPLIT by EffectiveDate."""
        current = events

        if IdentifierChangeType.RENAME in changeTypesToApply:
            current = self._applyRename(current)

        relsForSplits = [
            r for r in current.IdentifierRelationships if isinstance(r, IdentifierRelationshipSplit) and r.ChangeType in changeTypesToApply
        ]
        for rel in sorted(relsForSplits, key=lambda r: r.EffectiveDate):
            current = self._applySplitOrReverseSplit(current, rel)

        return current

    def _applyRename(self, events: pfe.FinancialEvents) -> pfe.FinancialEvents:
        """Merge groupings by rename chains and normalize identifiers to the sink (proper) identifier."""
        rels = [r for r in events.IdentifierRelationships if r.ChangeType == IdentifierChangeType.RENAME]
        if not rels:
            return events

        pairs = [(r.FromIdentifier, r.ToIdentifier) for r in rels]

        def _sink(identifier: FinancialIdentifier) -> FinancialIdentifier:
            """Follow rename edges to the end of the chain (sink)."""
            for from_id, to_id in pairs:
                if from_id.isTheSameAs(identifier):
                    return _sink(to_id)
            return identifier

        # Every identifier in any relationship (from or to) -> its chain sink. FinancialIdentifier is hashable, so dict works.
        id_to_sink: dict[FinancialIdentifier, FinancialIdentifier] = {}
        for r in rels:
            id_to_sink[r.FromIdentifier] = _sink(r.FromIdentifier)
            id_to_sink[r.ToIdentifier] = _sink(r.ToIdentifier)

        def _getSink(identifier: FinancialIdentifier) -> FinancialIdentifier | None:
            for node, sink in id_to_sink.items():
                if node.sameInstrumentByIsin(identifier) or sink.sameInstrumentByIsin(identifier):
                    return sink
            return None

        notAffected: list[pgf.FinancialGrouping] = []
        by_sink: dict[FinancialIdentifier, list[pgf.FinancialGrouping]] = {}
        for g in events.Groupings:
            sink = _getSink(g.FinancialIdentifier)
            if sink is None:
                notAffected.append(g)
            else:
                by_sink.setdefault(sink, []).append(g)

        merged = [self._mergeGroupings(sink, group_list) for sink, group_list in by_sink.items()]
        return pfe.FinancialEvents(
            Groupings=notAffected + merged,
            IdentifierRelationships=events.IdentifierRelationships,
        )

    def _applySplitOrReverseSplit(
        self,
        events: pfe.FinancialEvents,
        relationship: IdentifierRelationshipSplit,
    ) -> pfe.FinancialEvents:
        """
        Apply one SPLIT or REVERSE_SPLIT: scale pre-effective-date quantities in the grouping
        for the old identifier (FromIdentifier), then merge that grouping into the one for the
        new identifier (ToIdentifier). Result: a single grouping under the new ISIN with scaled
        historical quantities and any post-split activity already under that ISIN.
        """
        if relationship.QuantityBefore == 0:
            return events
        ratio = relationship.QuantityAfter / relationship.QuantityBefore
        from_grouping = next(
            (g for g in events.Groupings if relationship.FromIdentifier.sameInstrumentByIsin(g.FinancialIdentifier)),
            None,
        )
        if from_grouping is None:
            return events

        scaled_trades: list[pe.TradeEventStockAcquired | pe.TradeEventStockSold] = []
        for t in from_grouping.StockTrades:
            if not t.Date < relationship.EffectiveDate:
                scaled_trades.append(t)
                continue

            m = t.ExchangedMoney
            new_money = cf.GenericMonetaryExchangeInformation(
                UnderlyingCurrency=m.UnderlyingCurrency,
                UnderlyingQuantity=m.UnderlyingQuantity * ratio,
                UnderlyingTradePrice=m.UnderlyingTradePrice * 1 / ratio,
                ComissionCurrency=m.ComissionCurrency,
                ComissionTotal=m.ComissionTotal,
                TaxCurrency=m.TaxCurrency,
                TaxTotal=m.TaxTotal,
                FxRateToBase=m.FxRateToBase,
            )

            if isinstance(t, pe.TradeEventStockAcquired):
                scaled_trades.append(
                    pe.TradeEventStockAcquired(
                        ID=t.ID,
                        FinancialIdentifier=t.FinancialIdentifier,
                        AssetClass=t.AssetClass,
                        Date=t.Date,
                        Multiplier=t.Multiplier,
                        ExchangedMoney=new_money,
                        AcquiredReason=t.AcquiredReason,
                    )
                )
            else:
                scaled_trades.append(
                    pe.TradeEventStockSold(
                        ID=t.ID,
                        FinancialIdentifier=t.FinancialIdentifier,
                        AssetClass=t.AssetClass,
                        Date=t.Date,
                        Multiplier=t.Multiplier,
                        ExchangedMoney=new_money,
                    )
                )

        scaled_lots: list[pgf.TaxLotStock] = []
        for lot in from_grouping.StockTaxLots:
            if lot.Acquired.Date < relationship.EffectiveDate:
                scaled_lots.append(
                    pgf.TaxLotStock(
                        ID=lot.ID,
                        FinancialIdentifier=lot.FinancialIdentifier,
                        Quantity=lot.Quantity * ratio,
                        Acquired=lot.Acquired,
                        Sold=lot.Sold,
                        ShortLongType=lot.ShortLongType,
                    )
                )
            else:
                scaled_lots.append(lot)

        scaled_from = pgf.FinancialGrouping(
            FinancialIdentifier=from_grouping.FinancialIdentifier,
            CountryOfOrigin=from_grouping.CountryOfOrigin,
            UnderlyingCategory=from_grouping.UnderlyingCategory,
            StockTrades=scaled_trades,
            StockTaxLots=scaled_lots,
            DerivativeGroupings=from_grouping.DerivativeGroupings,
            CashTransactions=from_grouping.CashTransactions,
        )

        to_grouping = next(
            (g for g in events.Groupings if relationship.ToIdentifier.sameInstrumentByIsin(g.FinancialIdentifier)),
            None,
        )
        if to_grouping is not None:
            merged = self._mergeGroupings(relationship.ToIdentifier, [scaled_from, to_grouping])
        else:
            merged = self._mergeGroupings(relationship.ToIdentifier, [scaled_from])

        other = [g for g in events.Groupings if g is not from_grouping and g is not to_grouping]
        return pfe.FinancialEvents(
            Groupings=other + [merged],
            IdentifierRelationships=events.IdentifierRelationships,
        )

    def _mergeGroupings(
        self,
        properId: FinancialIdentifier,
        groupings: list[pgf.FinancialGrouping],
    ) -> pgf.FinancialGrouping:
        first = groupings[0]
        allStockTrades: list[pe.TradeEventStockAcquired | pe.TradeEventStockSold] = []
        allStockTaxLots: list[pgf.TaxLotStock] = []
        allDerivativeGroupings: list[pgf.DerivativeGrouping] = []
        allCashTransactions: list[pe.TransactionCash] = []

        for g in groupings:
            for t in g.StockTrades:
                allStockTrades.append(self._withIdentifier(t, properId))
            for lot in g.StockTaxLots:
                allStockTaxLots.append(self._lotWithIdentifier(lot, properId))
            for dg in g.DerivativeGroupings:
                allDerivativeGroupings.append(self._derivativeGroupingWithIdentifier(dg, properId))
            for c in g.CashTransactions:
                allCashTransactions.append(self._withIdentifier(c, properId))

        return pgf.FinancialGrouping(
            FinancialIdentifier=properId,
            CountryOfOrigin=first.CountryOfOrigin,
            UnderlyingCategory=first.UnderlyingCategory,
            StockTrades=allStockTrades,
            StockTaxLots=allStockTaxLots,
            DerivativeGroupings=allDerivativeGroupings,
            CashTransactions=allCashTransactions,
        )

    def _withIdentifier(self, event: pe.TradeEvent, identifier: FinancialIdentifier) -> pe.TradeEvent:
        event.FinancialIdentifier = identifier
        return event

    def _lotWithIdentifier(self, lot: pgf.TaxLotStock, identifier: FinancialIdentifier) -> pgf.TaxLotStock:
        return pgf.TaxLotStock(
            ID=lot.ID,
            FinancialIdentifier=identifier,
            Quantity=lot.Quantity,
            Acquired=self._withIdentifier(lot.Acquired, identifier),
            Sold=self._withIdentifier(lot.Sold, identifier),
            ShortLongType=lot.ShortLongType,
        )

    def _derivativeGroupingWithIdentifier(
        self,
        dg: pgf.DerivativeGrouping,
        identifier: FinancialIdentifier,
    ) -> pgf.DerivativeGrouping:
        newTrades = [self._withIdentifier(t, identifier) for t in dg.DerivativeTrades]

        newLots = [
            pgf.TaxLotDerivative(
                ID=lot.ID,
                FinancialIdentifier=identifier,
                Quantity=lot.Quantity,
                Acquired=self._withIdentifier(lot.Acquired, identifier),
                Sold=self._withIdentifier(lot.Sold, identifier),
                ShortLongType=lot.ShortLongType,
            )
            for lot in dg.DerivativeTaxLots
        ]

        return pgf.DerivativeGrouping(
            FinancialIdentifier=identifier,
            DerivativeTrades=newTrades,
            DerivativeTaxLots=newLots,
        )

File: lib/core/financialEvents/ApplyIdentifierRelationshipsService.ts
import type { TradeEventStockAcquired, TradeEventStockSold, TransactionCash } from "@brrr/core/schemas/Events.ts";
import type { FinancialEvents } from "@brrr/core/schemas/FinancialEvents.ts";
import type { FinancialIdentifier } from "@brrr/core/schemas/FinancialIdentifier.ts";
import type { DerivativeGrouping, FinancialGrouping } from "@brrr/core/schemas/Grouping.ts";
import type { IdentifierRelationshipSplit } from "@brrr/core/schemas/IdentifierRelationship.ts";
import { IdentifierChangeType } from "@brrr/core/schemas/IdentifierRelationship.ts";
import type { TaxLotStock } from "@brrr/core/schemas/Lots.ts";
import type { AnyProvenanceStep, RenameProvenanceStep, SplitProvenanceStep } from "@brrr/core/schemas/Provenance.ts";

export class ApplyIdentifierRelationshipsService {
	apply(events: FinancialEvents, changeTypesToApply: IdentifierChangeType[]): FinancialEvents {
		let current = events;

		if (changeTypesToApply.includes(IdentifierChangeType.RENAME)) {
			current = this._applyRename(current);
		}

		const relsForSplits = current.identifierRelationships.filter(
			(r): r is IdentifierRelationshipSplit => "quantityBefore" in r && changeTypesToApply.includes(r.changeType),
		);

		for (const rel of [...relsForSplits].sort((a, b) => a.effectiveDate.toMillis() - b.effectiveDate.toMillis())) {
			current = this._applySplitOrReverseSplit(current, rel);
		}

		return current;
	}

	private _applyRename(events: FinancialEvents): FinancialEvents {
		const rels = events.identifierRelationships.filter((r) => r.changeType === IdentifierChangeType.RENAME);
		if (rels.length === 0) return events;

		const pairs = rels.map((r) => [r.fromIdentifier, r.toIdentifier] as const);

		const _sink = (identifier: FinancialIdentifier): FinancialIdentifier => {
			for (const [fromId, toId] of pairs) {
				if (fromId.isTheSameAs(identifier)) return _sink(toId);
			}
			return identifier;
		};

		const idToSink = new Map<FinancialIdentifier, FinancialIdentifier>();
		for (const r of rels) {
			idToSink.set(r.fromIdentifier, _sink(r.fromIdentifier));
			idToSink.set(r.toIdentifier, _sink(r.toIdentifier));
		}

		const _getSink = (identifier: FinancialIdentifier): FinancialIdentifier | null => {
			for (const [node, sink] of idToSink) {
				if (node.sameInstrumentByIsin(identifier) || sink.sameInstrumentByIsin(identifier)) {
					return sink;
				}
			}
			return null;
		};

		const notAffected: FinancialGrouping[] = [];
		const bySink = new Map<string, { sink: FinancialIdentifier; groupings: FinancialGrouping[] }>();

		for (const g of events.groupings) {
			const sink = _getSink(g.financialIdentifier);
			if (sink === null) {
				notAffected.push(g);
			} else {
				const key = sink.toKey();
				const existing = bySink.get(key);
				if (existing) {
					existing.groupings.push(g);
				} else {
					bySink.set(key, { sink, groupings: [g] });
				}
			}
		}

		// Build provenance per sink
		const sinkToProvenance = new Map<string, AnyProvenanceStep[]>();
		for (const r of rels) {
			const sink = _sink(r.toIdentifier);
			const key = sink.toKey();
			const step: RenameProvenanceStep = {
				kind: "rename",
				fromIdentifier: r.fromIdentifier,
				toIdentifier: r.toIdentifier,
				changeType: r.changeType,
				effectiveDate: r.effectiveDate,
			};
			const existing = sinkToProvenance.get(key) ?? [];
			// Avoid duplicates
			if (!existing.some((s) => s.fromIdentifier.isTheSameAs(step.fromIdentifier) && s.toIdentifier.isTheSameAs(step.toIdentifier))) {
				existing.push(step);
			}
			sinkToProvenance.set(key, existing);
		}

		const merged = [...bySink.values()].map(({ sink, groupings }) =>
			this._mergeGroupings(sink, groupings, sinkToProvenance.get(sink.toKey()) ?? [])
		);

		return {
			groupings: [...notAffected, ...merged],
			identifierRelationships: events.identifierRelationships,
		};
	}

	private _applySplitOrReverseSplit(events: FinancialEvents, relationship: IdentifierRelationshipSplit): FinancialEvents {
		if (relationship.quantityBefore === 0) return events;
		const ratio = relationship.quantityAfter / relationship.quantityBefore;

		const fromGrouping = events.groupings.find((g) => relationship.fromIdentifier.sameInstrumentByIsin(g.financialIdentifier)) ?? null;
		if (!fromGrouping) return events;

		const scaledTrades: (TradeEventStockAcquired | TradeEventStockSold)[] = [];
		for (const t of fromGrouping.stockTrades) {
			if (!(t.date < relationship.effectiveDate)) {
				scaledTrades.push(t);
				continue;
			}

			const m = t.exchangedMoney;
			const step: SplitProvenanceStep = {
				kind: "split",
				fromIdentifier: relationship.fromIdentifier,
				toIdentifier: relationship.toIdentifier,
				changeType: relationship.changeType,
				effectiveDate: relationship.effectiveDate,
				quantityBefore: m.underlyingQuantity,
				quantityAfter: m.underlyingQuantity * ratio,
				beforeQuantity: m.underlyingQuantity,
				beforeTradePrice: m.underlyingTradePrice,
				beforeExchangedMoney: m,
			};

			const newMoney = {
				...m,
				underlyingQuantity: m.underlyingQuantity * ratio,
				underlyingTradePrice: m.underlyingTradePrice * (1 / ratio),
			};

			if (t.kind === "StockAcquired") {
				scaledTrades.push({
					...t,
					exchangedMoney: newMoney,
					provenance: [...t.provenance, step],
				});
			} else {
				scaledTrades.push({
					...t,
					exchangedMoney: newMoney,
					provenance: [...t.provenance, step],
				});
			}
		}

		const scaledLots: TaxLotStock[] = [];
		for (const lot of fromGrouping.stockTaxLots) {
			if (lot.acquired.date < relationship.effectiveDate) {
				const step: SplitProvenanceStep = {
					kind: "split",
					fromIdentifier: relationship.fromIdentifier,
					toIdentifier: relationship.toIdentifier,
					changeType: relationship.changeType,
					effectiveDate: relationship.effectiveDate,
					quantityBefore: relationship.quantityBefore,
					quantityAfter: relationship.quantityAfter,
					beforeQuantity: lot.quantity,
				};
				scaledLots.push({
					...lot,
					quantity: lot.quantity * ratio,
					provenance: [...lot.provenance, step],
				});
			} else {
				scaledLots.push(lot);
			}
		}

		const splitStep: SplitProvenanceStep = {
			kind: "split",
			fromIdentifier: relationship.fromIdentifier,
			toIdentifier: relationship.toIdentifier,
			changeType: relationship.changeType,
			effectiveDate: relationship.effectiveDate,
			quantityBefore: relationship.quantityBefore,
			quantityAfter: relationship.quantityAfter,
		};

		const scaledFrom: FinancialGrouping = {
			...fromGrouping,
			stockTrades: scaledTrades,
			stockTaxLots: scaledLots,
			provenance: [...fromGrouping.provenance, splitStep],
		};

		const toGrouping =
			events.groupings.find((g) => g !== fromGrouping && relationship.toIdentifier.sameInstrumentByIsin(g.financialIdentifier)) ??
				null;

		let mergedGrouping: FinancialGrouping;
		if (toGrouping !== null) {
			mergedGrouping = this._mergeGroupings(
				relationship.toIdentifier,
				[scaledFrom, toGrouping],
				[...scaledFrom.provenance, ...toGrouping.provenance],
			);
		} else {
			mergedGrouping = this._mergeGroupings(
				relationship.toIdentifier,
				[scaledFrom],
				scaledFrom.provenance,
			);
		}

		const other = events.groupings.filter((g) => g !== fromGrouping && g !== toGrouping);
		return {
			groupings: [...other, mergedGrouping],
			identifierRelationships: events.identifierRelationships,
		};
	}

	private _mergeGroupings(
		properId: FinancialIdentifier,
		groupings: FinancialGrouping[],
		groupingProvenance: AnyProvenanceStep[],
	): FinancialGrouping {
		const first = groupings[0];
		const allStockTrades: (TradeEventStockAcquired | TradeEventStockSold)[] = [];
		const allStockTaxLots: TaxLotStock[] = [];
		const allDerivativeGroupings: DerivativeGrouping[] = [];
		const allCashTransactions: TransactionCash[] = [];

		for (const g of groupings) {
			for (const t of g.stockTrades) allStockTrades.push({ ...t, financialIdentifier: properId });
			for (const lot of g.stockTaxLots) {
				allStockTaxLots.push({
					...lot,
					financialIdentifier: properId,
					acquired: { ...lot.acquired, financialIdentifier: properId },
					sold: { ...lot.sold, financialIdentifier: properId },
				});
			}
			for (const dg of g.derivativeGroupings) {
				allDerivativeGroupings.push({
					...dg,
					financialIdentifier: properId,
					derivativeTrades: dg.derivativeTrades.map((t) => ({ ...t, financialIdentifier: properId })),
					derivativeTaxLots: dg.derivativeTaxLots.map((lot) => ({
						...lot,
						financialIdentifier: properId,
						acquired: { ...lot.acquired, financialIdentifier: properId },
						sold: { ...lot.sold, financialIdentifier: properId },
					})),
				});
			}
			for (const c of g.cashTransactions) allCashTransactions.push({ ...c, financialIdentifier: properId });
		}

		return {
			financialIdentifier: properId,
			countryOfOrigin: first.countryOfOrigin,
			underlyingCategory: first.underlyingCategory,
			stockTrades: allStockTrades,
			stockTaxLots: allStockTaxLots,
			derivativeGroupings: allDerivativeGroupings,
			cashTransactions: allCashTransactions,
			provenance: groupingProvenance,
		};
	}
}
